pennylane/ops/qubit.py

```diff
@@ -2165,6 +2165,39 @@ def _controlled(self, wire):
         ControlledQubitUnitary(*self.parameters, control_wires=wire, wires=self.wires)
 
 
+class QubitMatrix(Operation):
+    r"""QubitMatrix(A, wires)
+    Apply an arbitrary square matrix.
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 1
+    * Gradient recipe: None
+
+    Args:
+        A (array[complex]): square matrix
+        wires (Sequence[int] or int): the wire(s) the operation acts on
+    """
+
+    num_wires = AnyWires
+    num_params = 1
+    par_domain = "A"
+    grad_method = None
+
+    @classmethod
+    def _matrix(cls, *params):
+        A = np.asarray(params[0])
```

Comment:
@anthayes92 this would look like the following:

```python
@classmethod
def _matrix(cls, *params):
    A = params[0]
    shape = qml.math.shape(A)
    
    if shape[0] != shape[1]:
        raise ValueError("Operator must be a square matrix.")

    return A
```

This way, we are not making any assumptions as to the type of A, and we are avoiding casting it to other types.

Reply:
thanks for the example! Removed numpy calls here: 037fd9b79d6c67164a326cfe2a4175485fdebe2f